### PR TITLE
[Proposal] Organize spell scripts in folders, depending of type

### PR DESCRIPTION
**Pull request by Xaver-DaRed**
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [x] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

Tested and confirmed working. All spells working as they were before.

A proposal for a new spell organization, similar to how Blue Magic and Trusts are done. It follows in-game menu organization.
Things of note while this was done:
- We have 2 copies of Blue Spells out of bluemagic folder. They are also inside of bluemagic folder.
- We have 2 non-spell "global" scripts inside spells folder.
- We have a doom spell. It has no entry in database, unlike all other spells.

End result looks like this.
imagen

### CODE REVIEW

**Comment by claywar** [comment]
I'm onboard with this.  I would suggest renaming blackmagic, bluemagic and whitemagic just black, blue, white though for simplicity since they're a subdir in spells.

**Comment by TeoTwawki** [comment]
I'm on board with basically following the in game menu layout. diff menu selection = gets a diff directory path

There are a few spells that in retail are mob-only but are castable by players if set to be in the database (Curse comes to mind) we can reliably guess where these _would_ fit though.

We could potentially use the main dir as a fallback if a spell script isn't found elsewhere.

File: src/map/lua/luautils.cpp [hunk 1]
```diff
@@ -532,9 +532,51 @@ namespace luautils
 
         auto name = (const char*)PSpell->getName();
 
-        if (PSpell->getSpellGroup() == SPELLGROUP_BLUE)
+        if (PSpell->getSpellGroup() == SPELLGROUP_WHITE)
```
Comment (zach2good) [inline src/map/lua/luautils.cpp]: Please change this to a `switch` statement:
```cpp
std::string switchKey = "";
switch (PSpell->getSpellGroup())
{
    case SPELLGROUP_WHITE:
    {
        switchKey = "white";
    }
    break;
    ...
    default:
    {
        ShowError(...);
    }
    break;
}

if (auto cached_func = lua["xi"]["globals"]["spells"][switchKey][name][funcName]; cached_func.valid())
{
    return cached_func;
}
...
```

File: src/map/spell.cpp [hunk 2]
```diff
@@ -483,9 +483,33 @@ namespace spell
                 PSpellList[static_cast<uint16>(PSpell->getID())] = PSpell;
 
                 auto filename = fmt::format("./scripts/globals/spells/{}.lua", PSpell->getName());
-                if (PSpell->getSpellGroup() == SPELLGROUP_BLUE)
+                if (PSpell->getSpellGroup() == SPELLGROUP_WHITE)
```
Comment (zach2good) [inline src/map/spell.cpp]: Same here, but you can do the format:
```cpp
std::string switchKey = "";

// Switch

filename = fmt::format("./scripts/globals/spells/{}/{}.lua", switchKey, PSpell->getName());
```

**Review by zach2good — CHANGES_REQUESTED** [review]
Good idea, just needs some cleanup in core (I assume Lua is fine, I'm not going through 600+ files 👀)

It might be worth making a function in spellutils.cpp (or similar) to turn the `SPELLGROUP` enum into the desired string (using the switch I posted).

**Review by zach2good — APPROVED** [review]

**Pull request merged**